File: solution/c#/Day10/Day10/FizzBuzz.cs
namespace Day10
{
    public static class FizzBuzz
    {
        private const int Min = 0;
        private const int Max = 100;
        private const int Fizz = 3;
        private const int Buzz = 5;
        private const int Fizz_Buzz = 15;

        private static readonly IReadOnlyDictionary<Predicate<int>, Func<int, string>> mapping =
            new Dictionary<Predicate<int>, Func<int, string>>
            {
                {i => Is(Fizz_Buzz, i), i => "FizzBuzz"},
                {i => Is(Fizz, i), i => "Fizz"},
                {i => Is(Buzz, i), i => "Buzz"},
                {i => true, i => i.ToString()},
            };

        public static string Convert(int input)
        {
            var mappingFunction = mapping
                .Where(kvp => !IsOutOfRange(input))
                .Where(kvp => kvp.Key(input))
                .Select(kvp => kvp.Value)
                .FirstOrDefault();

            return mappingFunction != null
                ? mappingFunction(input)
                : throw new OutOfRangeException();
        }

        private static bool Is(int divisor, int input) => input % divisor == 0;

        private static bool IsOutOfRange(int input) => input is <= Min or > Max;
    }
}

File: exercise/ts/day14/src/fizzbuzz.ts
const min = 1;
const max = 100;
const fizz = 3;
const buzz = 5;
const fizzBuzz = 15;

let mapping: Map<(input: number) => boolean, (input: number) => string> = new Map([
    [(input: number) => is(fizzBuzz, input), () => "FizzBuzz"],
    [(input: number) => is(fizz, input), () => "Fizz"],
    [(input: number) => is(buzz, input), () => "Buzz"],
    [(input: number) => !isOutOfRange(input), (input: number) => input.toString()],
]);

export function fizzbuzz(input: number): string {
    switch (isOutOfRange(input)) {
        case true:
            throw new Error("Input is out of range");
        case false:
            return (firstMatchingPredicateFor(input))[1](input);
    }
}

const is = (divisor: number, input: number): boolean => input % divisor === 0;
const isOutOfRange = (input: number): boolean => input < min || input > max;
const firstMatchingPredicateFor = (input: number) =>
    Array.from(mapping.entries())
        .find(([predicate]) => predicate(input));